### PR TITLE
golang: provide method to refresh route cache

### LINKED ISSUE #36848: golang: add ability to refresh route cache once we change the request
*Title*: *golang: add ability to refresh route cache once we change the request*

*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

Sometimes, we want to change the matched route by modifying the header, but not decided by the final header status. For example, considering that the client's request path is `/disney/api`, and the path sent to the upstream is `/api/`, but we want to change the matched route as `/user/api/`:

```
		header.SetPath("/user/api/") // path used to match the new route
		f.callbacks.ClearRouteCache()
		header.SetPath("/api/") // path used by the upstream
```

For now, as Envoy will only refresh the route cache after decoding headers, the final header status `/api/` is used to match the new route.

Therefore, I propose adding a new method `RefreshRouteCache()`, which clears the route cache and rematch the route. With this new method:

```
		header.SetPath("/user/api/") // path used to match the new route
		f.callbacks.RefreshRouteCache()
		header.SetPath("/api/")
```

The `/user/api/` is used to match the new route.

CC @doujiang24 

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

### PR DESCRIPTION
**Pull request by spacewander**
Commit Message: golang: provide method to refresh route cache
Additional Description:
Risk Level: Low, add new method
Testing: Integration
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fixes #36848
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: contrib/golang/filters/http/source/golang_filter.cc
Comment (doujiang24) [inline contrib/golang/filters/http/source/golang_filter.cc:920]: Maybe it's better to add a bool param in `clearRouteCache(bool refresh)`? it may save some similar code.
Adding a new `RefreshRouteCache` Golang method is friendly for users.

**Review by doujiang24 — COMMENTED** [review]
Just a tiny comment, otherwise, lgtm

File: contrib/golang/filters/http/source/golang_filter.cc
Comment (spacewander) [reply contrib/golang/filters/http/source/golang_filter.cc]: @doujiang24 
Updated.

**Review by doujiang24 — APPROVED** [review]
lgtm

**Comment by doujiang24** [comment]
@nezdolik It's ready to merge, thanks.

/assign @nezdolik

**Comment by doujiang24** [comment]
It's ready to merge, please help to merge it, thanks~
cc @phlax @wbpcode

**Review by wbpcode — APPROVED** [review]

**Comment by wbpcode** [comment]
blindly approve because extension maintainer has given lgtm. :)

**Pull request merged**
